src/odoc/bin/main.ml

```diff
@@ -187,14 +192,34 @@ end = struct
               "Either --package or --parent should be specified, not both")
     in
     let source =
-      match (source_parent_file, source_name) with
-      | Some parent, Some name -> Ok (Some (parent, name, source_cmt))
-      | Some _, None | None, Some _ ->
+      match
+        (source_parent_file, source_name, source_cmt, Fs.File.get_ext input)
+      with
+      | Some parent, Some name, None, ".cmt" -> Ok (Some (parent, name, input))
+      | Some parent, Some name, Some cmt, ".cmt" ->
+          if Fpath.equal cmt input then Ok (Some (parent, name, input))
+          else
+            Error
+              (`Cli_error
+                "--cmt has to be equal to the input file when this one has \
+                 .cmt extension.")
+      | Some parent, Some name, Some cmt, _ -> Ok (Some (parent, name, cmt))
+      | Some _, Some _, None, _ ->
+          Error
+            (`Cli_error
+              "--cmt has to be passed when --source-parent-file and \
+               --source-name are passed and the input file is not a cmt file.")
+      | Some _, None, _, _ | None, Some _, _, _ ->
           Error
             (`Cli_error
               "--source-parent-file and --source-name must be passed at the \
                same time.")
-      | None, None -> Ok None
+      | None, None, Some _, _ ->
+          Error
+            (`Cli_error
+              "--cmt should only be passed when --source-parent-file and \
+               --source-name are passed.")
```

Comment:
Also, we might want to handle this case too - e.g. for autogenerated sources (ie, `foo.ml-gen`) - in which case you'll need the cmt to handle shape lookup but maybe you don't want the source rendered.

Reply:
I planned to do that with occurrence count (we might want to count occurrence but do not render the source).
So yes, we should do that. (even though I would prefer all sources rendered, including the auto-generated one).